### ISSUE TITLE
[rfc7233] If-Range: isRangeFresh bug?

### ISSUE THREAD

**Issue by silvioprog** [pillarjs/send#111]
Hello,

I'm testing the `send` module, and I think I found a bug. Consider the following code:

``` js
var express = require('express');
var app = express();

app.get('/', function (req, res, next) {
  res.sendFile('abc123.txt');
}).listen(8080, function () {
  console.log('Listening on port 8080');
});
```

The `abc123.txt` content:

```
abc123
```

And my request is:

```
GET: http://localhost:8080
Range: bytes=2-4
If-Range: W/"6-15521ebc6bc"
```

So, the result is:

```
c12
```

However, conforming rfc7233:

>  A client MUST NOT generate an If-Range header field containing an **entity-tag that is marked as weak**.

...

> ... A valid entity-tag can be distinguished from a valid HTTP-date **by examining the first two characters for a DQUOTE**.

But, it seems `isRangeFresh()` looks any ocurrence of `"` char instead of first ones:

``` js
...
[Line 382]
SendStream.prototype.isRangeFresh = function isRangeFresh(){
  var ifRange = this.req.headers['if-range'];

  if (!ifRange) return true;

  return ~ifRange.indexOf('"')
    ? ~ifRange.indexOf(this.res._headers['etag'])
    : Date.parse(this.res._headers['last-modified']) <= Date.parse(ifRange);
};
...
```

The same test on nginx works properly. The hash generated by nginx is `"575475b3-6"`, and when I send the `If-Range` with `W/"575475b3-6"`, it returns the entire content, conforming rfc.

**Comment by dougwilson** [comment]
Haha, yea, I noticed that when I was fixing a bunch of other range bugs (I didn't write the module, but maintain it now). Since neither our ETags nor the date are strong validators, to follow the RFC, this module should actually never respond with a ranged request if there is a If-Range header.

I didn't fix it with the rest of the range things, since this behavior has been this way for over 5 years already without any reported issue.

Besides you simply noticing that it does not technically conform, can you tell me if this is actually creating an issue with a particular web browser or other client?

**Comment by silvioprog** [comment]
Thanks for reply @dougwilson

> Haha, yea, I noticed that when I was fixing a bunch of other range bugs (I didn't write the module, but maintain it now). Since neither our ETags nor the date are strong validators, to follow the RFC, this module should actually never respond with a ranged request if there is a If-Range header.

I did a fix conforming rfc, after that, it works properly like nginx. The rfc doesn't says how the server treats an entity-tag that is marked as weak, so I kept it as is.

> I didn't fix it with the rest of the range things, since this behavior has been this way for over 5 years already without any reported issue.

Well, a bug may appear at any time and in any project, even in stable codes. :-)

> Besides you simply noticing that it does not technically conform, can you tell me if this is actually creating an issue with a particular web browser or other client?

I'm using the [Postman](https://www.getpostman.com/) on Google Chrome, but the bug is more related to the rfc conformity, so the problem hapens in other clients too, like `curl` by command line.

I have good news, please take a look at this attached patch, it fix the `isRangeFresh()` function and its respective test. Feel free to make any change, because my JavaScript knowledge is a little bit small, so this patch can be easily improved.

[0001-Fix-wrong-If-Range-parsing-111.-patch-by-silvioprog.patch.txt](https://github.com/pillarjs/send/files/310405/0001-Fix-wrong-If-Range-parsing-111.-patch-by-silvioprog.patch.txt)

**Comment by dougwilson** [comment]
Thanks for the patch, but it still does not comply with the RFC. Like I said, neither the etag nor the date are strong validators, so we cannot respond with a ranged request even with the date matching to follow the RFC.in addition,the test change you make does not make sense.

Let me know when you have a new patch to try, otherwise I'm not going to fix this unless you can demonstrate there is an actual issue and not just tat we are ignoring a MUST that, in essence, the client should probably not even be doing in the fist place.

**Comment by dougwilson** [comment]
Please user the github pull requests to offer patches so I can actually comment on aspects of the patch, if you would like to move this forward. And remember, if the point of the patch is to comply with the RFC, please make sure that is what you are actually doing; the given patch still does not comply with the RFC's requirements.

**Comment by dougwilson** [comment]
Darn close button being next to the comment button.

**Comment by dougwilson** [comment]
> > Besides you simply noticing that it does not technically conform, can you tell me if this is actually creating an issue with a particular web browser or other client?

> I'm using the Postman on Google Chrome, but the bug is more related to the rfc conformity, so the problem happens in other clients too, like  curl  by command line.

Sorry, you misunderstood my question. If you are adding those headers yourself, that does not found as a client issue, as you are manually adding those headers and hence is only an issue with you specifically noticing it. What I mean is that,over the course of using this module and a standard client interacting with the web server (i.e. when you are personally not involved in constructing the headers in the request), does this actually exhibit unexpected behavior in a web browser?

Based on your response, it sounds like the answer is "no" and if this is the case, i won't fix this until 1.0.0 of this module, which is unlikely to release for a few more years.

If you have a real example of a web browser or other client encountering an actual problem (like it causes a corrupted response, corrupted cache, something), please prove this by providing instruction i can use to reproduce and we can consider adding it before the next major version.

If you simply want to improve the conformance with the RFC, then please open a pull request with a patch. In order to comply with the RFC, the isRangeFresh function of our module simply must always return false. Be sure to add new tests as well!

If you are not very familiar with JavaScript, that is fine! I'd be happy to fix it, but only if there is a real client issue. If it is just for the RFC technicality,I do not have any interest in fixing it, and unless you can get someone else to help make the patch for you, you're welcome to close this issue.

**Comment by dougwilson** [comment]
I'm going to close this since there hasn't been any follow up. Please feel free to short over a PR if you're up to it :)
